Lib9c/Delegation/RewardBase.cs

```diff
@@ -0,0 +1,188 @@
+    {
+        private const string StateTypeName = "reward_base";
+        private const long StateVersion = 1;
+        private readonly IComparer<Currency> _currencyComparer = new CurrencyComparer();
+
+        public RewardBase(
+            Address address,
+            BigInteger totalShares,
+            IEnumerable<Currency> currencies)
+            : this(
+                  address,
+                  totalShares,
+                  currencies.Select(c => c * 0),
+                  RecommendedSigFig(totalShares))
+        {
+        }
+
+        public RewardBase(Address address, IValue bencoded)
+            : this(address, (List)bencoded)
+        {
+        }
+
+        public RewardBase(
+            Address address,
+            BigInteger totalShares,
+            IEnumerable<FungibleAssetValue> rewardPortion,
+            int sigfig)
+        {
+            Address = address;
+
+            if (totalShares.Sign <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(totalShares));
+            }
+
+            TotalShares = totalShares;
+
+            if (!rewardPortion.Select(f => f.Currency).All(new HashSet<Currency>().Add))
+            {
+                throw new ArgumentException("Duplicated currency in reward base.");
+            }
+
+            RewardPortion = rewardPortion.ToImmutableDictionary(f => f.Currency, f => f);
+            SigFig = sigfig;
+        }
+
+
+        public RewardBase(Address address, List bencoded)
+        {
+            if (bencoded[0] is not Text text || text != StateTypeName || bencoded[1] is not Integer integer)
+            {
+                throw new InvalidCastException();
+            }
+
+            if (integer > StateVersion)
+            {
+                throw new FailedLoadStateException("Un-deserializable state.");
+            }
+
+            Address = address;
+            TotalShares = (Integer)bencoded[2];
+            var rewardPortion = ((List)bencoded[3]).Select(v => new FungibleAssetValue(v));
+
+            if (!rewardPortion.Select(f => f.Currency).All(new HashSet<Currency>().Add))
+            {
+                throw new ArgumentException("Duplicated currency in reward base.");
+            }
+
+            RewardPortion = rewardPortion.ToImmutableDictionary(f => f.Currency, f => f);
+            SigFig = (Integer)bencoded[4];
+        }
+
+        private RewardBase(
+            Address address,
+            BigInteger totalShares,
+            ImmutableDictionary<Currency, FungibleAssetValue> rewardPortion,
+            int sigfig)
+        {
+            Address = address;
+            TotalShares = totalShares;
+            RewardPortion = rewardPortion;
+            SigFig = sigfig;
+        }
+
+        public Address Address { get; }
+
+        public BigInteger TotalShares { get; }
+
+        public int SigFig { get; private set; }
+
+        public static int Margin => 2;
```

Comment:
Why is this margin needed?

Reply:
To secure precision.
When total delegation is 1000 and reward is 1,
We can secure 0.001, and it's enough.
But when total delegation is 1234 and reward is 1,
The actual value is 0.00081037...,
but we cannot secure 0.001 though.

Reply:
It looks better to define it at the top as a public constant rather than a property.